File: codes/python/p026.py
# 
# Solution to Project Euler problem 26
# Copyright (c) Sankha Saha. All rights reserved.
# 
#  
#   
# 

import itertools


def compute():
	ans = max(range(1, 1000), key=reciprocal_cycle_len)
	return str(ans)


def reciprocal_cycle_len(n):
	seen = {}
	x = 1
	for i in itertools.count():
		if x in seen:
			return i - seen[x]
		else:
			seen[x] = i
			x = x * 10 % n


if __name__ == "__main__":
	print(compute())

File: codes/java/p026.java
/* 
 * Solution to Project Euler problem 26
 * Copyright (c) Sankha Saha. All rights reserved.
 * 
 *  
 *   
 */

import java.util.HashMap;
import java.util.Map;


public final class p026 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p026().run());
	}
	
	
	public String run() {
		int bestNumber = 0;
		int bestLength = 0;
		for (int i = 1; i <= 1000; i++) {
			int len = getCycleLength(i);
			if (len > bestLength) {
				bestNumber = i;
				bestLength = len;
			}
		}
		return Integer.toString(bestNumber);
	}
	
	
	private static int getCycleLength(int n) {
		Map<Integer,Integer> stateToIter = new HashMap<>();
		int state = 1;
		for (int iter = 0; ; iter++) {
			if (stateToIter.containsKey(state))
				return iter - stateToIter.get(state);
			else {
				stateToIter.put(state, iter);
				state = state * 10 % n;
			}
		}
	}
	
}
